src/System.Private.CoreLib/shared/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -80,7 +80,7 @@ public ValueTaskSourceStatus GetStatus(short token)
         {
             ValidateToken(token);
             return
-                !_completed ? ValueTaskSourceStatus.Pending :
+                Volatile.Read(ref _continuation) == null || !_completed ? ValueTaskSourceStatus.Pending :
```

Comment:
Curious - is `Volatile` needed here?  Do we rely on `_continuation` and `_completed` being set in a particular order?
It feels like the order does not matter. Even swapping operands of `||` would seem to result in correct behavior.

Unless there is something subtle that relies on ordering of reads... In that case, perhaps a comment/hint could be added.

Reply:
Basically - I see how reading `_continuation` first could be helpful when ordering is free, but there seems to be no need to force that.

Reply:
I was thinking about reads of continuation moving earlier, before the IsCompleted check. But in reality the worst that happens should be that it returns as Pending when it could otherwise not, which would just need hooking up a continuation when it want strictly necessary. So I can remove it.

Reply:
Yes, that what I guessed. The ordering race is fairly unlikely and just reporting `Pending` is safe, since that is what we would return if we are called just a tiny bit earlier.
Thanks!

---
In reply to: [258315542](https://github.com/dotnet/coreclr/pull/22711#discussion_r258315542) [](ancestors = 258315542)